src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -648,6 +661,69 @@ public byte[] Http2AltSvcOriginUri
             return await GetHttpConnectionAsync(request, async, cancellationToken).ConfigureAwait(false);
         }
 
+        private bool IsConnectionMaximumReached(Http2Connection[]? currentHttp2Connections) => currentHttp2Connections != null && EnableMultipleHttp2Connections && currentHttp2Connections.Length == _poolManager.Settings._maxHttp2ConnectionsPerServer;
+
+        private Http2Connection PickRandomConnection(HttpRequestMessage request, Http2Connection[] currentHttp2Connections)
+        {
+            // All connections are busy so pick a random one to wait on it.
```

Comment:
Why did we decide this is ok? This is one of the main concerns I raised previously.

Reply:
Due to a high complexity of the other approach. @geoffkizer raised the concern that other solution is so complex that it will hinder other HTTP/2 optimization efforts. We still can go back to a more advanced version later, I won't delete that another branch.

Reply:
cc: @karelz

Reply:
I don't think we explicitly decided it was ok. I agree, it's an issue.

The problem is, given this approach, it's hard to implement this the correct way. We don't have a notification from the Http2Connection that a new stream has become available. We could add this, of course, and we could add some queuing logic (similar to HTTP/1.1) to queue requests when all connections are busy, etc, etc.

That said, I'm not sure it's worth the trouble. The key scenario here is to support multiple HTTP2 connections. My suggestion is that we remove the connection limit here. Meaning, when you enable multiple HTTP2 connections, the number of connections is unbounded. This is what WinHTTP does as well.

Reply:
That notification and queueing are implemented in my other PR #38748, but it added noticeable complexity as you can see. It still might be there is a way to achieve the same with some simpler mechanism, but I currently don't see how.

Reply:
- Connection notifies about an available stream via [StreamSlotAvailable](https://github.com/alnikola/runtime/blob/557decd28f4fac30771808f8cfeab4941aa9c548/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L809)
- [Requests are queued while waiting for a free stream](https://github.com/alnikola/runtime/blob/557decd28f4fac30771808f8cfeab4941aa9c548/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L758) thanks to built-in support in SemaphoreSlim